### PR TITLE
Projection from 2D -> 1D for common lines

### LINKED ISSUE #65: 2D to 1D projection for common lines
It would be great to have a similar API to 3D -> 2D projection for 2D -> 1D

cc @McHaillet

### PR DESCRIPTION
**Pull request by McHaillet**
@alisterburt I started some work on a PR for the 2D -> 1D projection. I copied and adapted the code your wrote for the real space 3D->2D projection to create an equivalent function. It would be nice if you could give some feedback on that code and where to place it. I now added it to project_real but some other libraries seems to have separate files for 3d/2d, should that be the case for this as well?

The same still needs to be done for Fourier space 2D->1D but that seems more complicated as besides functions in project_fourier, it also requires a sample_dft_2d in interpolation. 

I updated the coarse alignment example as well to test if the proper tilt-axis angle could be optimised based on common lines. Seems to work quite nicely! I start with an angle that is way off (-30, while the actual angle is -88.7), and it finds the correct angle back. Importantly, I need to run it on the raw tilt-series. If I first apply the xcorr shifts, the tilt-axis is quite off (-100).

This is a PR that closes #65 (partly).

### CODE REVIEW

File: src/libtilt/projection/__init__.py
Comment (alisterburt) [inline src/libtilt/projection/__init__.py:2]: something about this doesn't feel 100% clear - how about `project_volume` and `project_image`? 🙂

File: src/libtilt/projection/project_real.py
Comment (alisterburt) [inline src/libtilt/projection/project_real.py:130]:
```suggestion
        rotated_coordinates = torch.flip(rotated_coordinates, dims=(-1,))  # yx -> xy
```

think this is correct, can you confirm?

File: src/libtilt/projection/project_real.py
Comment (alisterburt) [inline src/libtilt/projection/project_real.py:137]:
```suggestion
            mode='bilinear',
```

Comment (alisterburt) [reply src/libtilt/projection/project_real.py:137]: no need to clarify volumetric API here

File: src/libtilt/projection/project_real.py
Comment (alisterburt) [inline src/libtilt/projection/project_real.py:93]:
```suggestion
    4. summing samples along `h` dimension of a `(h, w)` image.
```

you sum over `h` later

File: examples/coarse_alignment.py
Comment (alisterburt) [inline examples/coarse_alignment.py:130]: huh, do you agree this feels unexpected? Do you think it's a wraparound thing?

**Review by alisterburt — COMMENTED** [review]
@McHaillet once again, so so psyched to see this!!

I've left a few minor comments but very happy to merge this once those are addressed

File: examples/coarse_alignment.py
Comment (alisterburt) [inline examples/coarse_alignment.py:118]: I'm wondering if this is the right form for the loss function, I see how it should behave and it seems to optimise which is good! Typically we would minimise the sum of squared differences between all projections, is this way better? if so, how come? :-)

Given that the projections are small maybe something like this, starting from projections with shape `b, d`

```python
squared_differences = (projections - einops.rearrange(projections, 'b d -> b 1 d')) ** 2
loss = einops.reduce(squared_differences, 'b1 b2 d -> 1', reduction='sum')
```

Some duplicate work but the vectorisation should overall speed it up 🤷🏼‍♂️

**Comment by alisterburt** [comment]
> The same still needs to be done for Fourier space 2D->1D but that seems more complicated as besides functions in project_fourier, it also requires a sample_dft_2d in interpolation.

Good point! I can add this

File: src/libtilt/projection/project_real.py
Comment (McHaillet) [reply src/libtilt/projection/project_real.py]: you're right! and for the volume also updated

File: src/libtilt/projection/project_real.py
Comment (McHaillet) [reply src/libtilt/projection/project_real.py]: check

File: src/libtilt/projection/__init__.py
Comment (McHaillet) [reply src/libtilt/projection/__init__.py]: I changed it now to project_volume_real and project_image_real to not mess with potential variants from project_fourier. Let me know if you agree.

File: examples/coarse_alignment.py
Comment (McHaillet) [reply examples/coarse_alignment.py]: Good point, that is much nicer :) Updated now

File: examples/coarse_alignment.py
Comment (McHaillet) [reply examples/coarse_alignment.py]: Absolutely and I couldn't let it rest, so managed to come up with a fix (see message below).

**Comment by McHaillet** [comment]
I managed to make the optimization work on the coarse aligned tilt-series. Some images here of the 2D -> 1D projections after optimizing the tilt-axis angle. (each row is a projected tilt image)

Firstly this was the set of lines when optimizing on the 'raw' tilt-series:
optimized_common_lines_no_alignment

These were the lines when optimizing on the aligned tilt-series:
optimized_common_lines_with_alignment

On these I noticed that the right and left sides get some artifacts which come from the fact that the tilt image is masked before projecting. The line is then normalized to mean0std1, which creates artifacts from the mask.

After some experimenting, I thought to do a 2D->1D projection of the mask to correctly weight the information in the lines. This manages to make the axis angle optimization work also after applying the shifts :smile: 

optimized_lines_mask_weights

About the wrap-around thing, it seems to work here now. But still likely that it can cause some issues.

**Review by alisterburt — APPROVED** [review]
You're a wizard, those common line images are gorgeous 😍 great example of why vis is so important!

I had another thought when rereading the code - the 'grid' with one grid point is useful only if we might later want to increase the resolution to model the tilt axis varying over time... otherwise a bare torch parameter would look more normal

Amazing work once again! Let's merge and iterate here ☺️

**Comment by McHaillet** [comment]
Thanks!!

> I had another thought when rereading the code - the 'grid' with one grid point is useful only if we might later want to increase the resolution to model the tilt axis varying over time... otherwise a bare torch parameter would look more normal

I left it because I tried optimizing with multiple grid points that vary with tilt angle, which is described in the AreTomo paper. Although in my experience running AreTomo it only ever optimizes a single tilt-axis (in-plane rotation). But I agree, varying over time could make sense!

**Pull request merged**
